src/wallet/walletdb.h

```diff
@@ -314,6 +314,69 @@ class ZcashdUnifiedAddressMetadata {
     }
 };
 
+// Serialization wrapper for reading and writing RecipientAddress
+// as a pair of typecode and address bytes, similar to how unified address
+// receivers are written (but excluding the unknown receiver case)
+class CSerializeRecipientAddress {
+    libzcash::RecipientAddress recipient;
+    libzcash::ReceiverType typecode;
+    CSerializeRecipientAddress() {} // for serialization only
+
+    public:
+        CSerializeRecipientAddress(libzcash::RecipientAddress recipient): recipient(recipient) {}
+
+        template<typename Stream>
+        void Serialize(Stream& s) const {
+            std::visit(match {
+                [&](const CKeyID& keyId) {
+                    ReceiverTypeSer(libzcash::ReceiverType::P2PKH).Serialize(s);
+                    s << keyId;
+                },
+                [&](const CScriptID& scriptId) {
+                    ReceiverTypeSer(libzcash::ReceiverType::P2SH).Serialize(s);
+                    s << scriptId;
+                },
+                [&](const libzcash::SaplingPaymentAddress& saplingAddr) {
+                    ReceiverTypeSer(libzcash::ReceiverType::Sapling).Serialize(s);
+                    s << saplingAddr;
+                }
+            }, recipient);
+        }
+
+        template<typename Stream>
+        void Unserialize(Stream& s) {
+            // This cast is fine because ZIP 316 uses CompactSize serialization including the
+            // size limit, which means it is at most a uint32_t.
```

Comment:
Why does `ReadCompactSize` even return a `uint64_t` when `MAX_SIZE` fits in `uint32_t`? Oh well, sigh.